lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -848,19 +848,32 @@ joinStakePool ctx spl (ApiT poolId) (ApiT wid) passwd = do
     liftE1 = throwE . ErrSelectForDelegationNoSuchWallet
     liftE2 = throwE . ErrSignDelegationNoSuchWallet
     liftE3 = throwE . ErrSubmitTxNoSuchWallet
-    poolIsUnknown =
-        isNothing . L.find (\(ApiStakePool pId _ _ _) -> pId == ApiT poolId)
+
+poolIsUnknown :: PoolId -> [ApiStakePool] -> Bool
+poolIsUnknown poolId =
+    isNothing . L.find (\(ApiStakePool pId _ _ _) -> pId == ApiT poolId)
 
 quitStakePool
-    :: forall n k.
+    :: forall ctx s t n k.
         ( DelegationAddress n k
+        , Buildable (ErrValidateSelection t)
+        , s ~ SeqState n k
         , k ~ ShelleyKey
+        , HardDerivation k
+        , ctx ~ ApiLayer s t k
         )
-    => ApiT PoolId
+    => ctx
+    -> StakePoolLayer IO
+    -> ApiT PoolId
     -> ApiT WalletId
     -> ApiWalletPassphrase
     -> Handler (ApiTransaction n)
-quitStakePool _ _ _ = throwError err501
+quitStakePool ctx spl (ApiT poolId) (ApiT wid) pwd = do
+    (_, walMeta, _) <- liftHandler $ withWorkerCtx ctx wid throwE $
+        \wrk -> W.readWallet wrk wid
+    when (walMeta ^. #delegation == W.NotDelegating) $
+        liftHandler $ throwE ErrSelectForDelegationNotDelegating
+    actionStakePool ctx spl Quit (ApiT poolId) (ApiT wid) pwd
```

Comment:
**[change required]**

Do not generalize too early here. For now, leave `quitStakePool` and `joinStakePool` as two separate functions, even if it induces a bit of duplication regarding the error handling.

**Why**

Not everything that applies to `joinStakePool` applies to `quitStakePool` in term of validation. For instance, the first check:

```
   when (walMeta ^. #delegation == W.Delegating poolId) $
        liftHandler $ throwE (ErrSelectForDelegationPoolAlreadyJoined poolId)
```

Is actually wrong for `quitStakePool` because this is precisely what we want to enforce in `quitStakePool` !
It'd suggest to start simple, and simple keep them as separate functions, and then, later, identify what part can potentially be factored out.

Reply:
you are right, I was too much expansive here. Made joinStakePool and quitStakePool separate... and added `WrongPool` error